wordpress-mobile/WordPress-iOS
# Show a toast message (notice) when moving posts to trash

For context, this change is initiated from this conversation: p1694476848843439-slack-C0180B5PRJ4

## Changes

This PR changes how to present an "undo" action to users when moving a post to trash. Here is a video preview of before & after the change.

| Before (with an undo row)  | After (with a toast message) |
| ------------- | ------------- |
| <video src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1101828/7b781e26-cab4-4401-834b-f4561122fe43"></video>  | <video src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1101828/1df59f59-c018-4b86-b25f-488db2029931"></video>  |

### Problems of the "Undo" row

From design perspective, there are a couple of issues:
1. This feature here is the only one in the app that shows an "undo" row. In other actions in the posts list and other areas in the app, a toast message with an action button is commonly used.
2. After moving many posts to trash, the trashed posts' corresponding "undo" row has exactly the same appearance which is not intuitive to users when they do want to undo the previous trash action.

From implementation perspective, it's not straight-forward to implement keeping the trashed post position and placing a "undo" row in it. From a quick testing, the existing implementation has an obvious bug when trashing posts from bottom to up.

### New flow

As you can see in the video, the new design is showing a toast message with an "Undo" button, which is consistent with other actions like restoring post from trash.

### Implementation details

There are three parts involved in this PR

#### 1. Removing the `restorableStatus` stored property from `AbstractPost`

This property is used to kept the original post status when it's moved to trash. 

One obvious issue is, this property is not a "managed" property, which means it's not stored into the Core Data database. So, its value is guaranteed to be lost if multiple context objects are used to write and read this information.

Also, this property gives a false sense of a trashed post can be "restored" to _any_ status, but I believe in reality a trashed post can only be restored to draft.

#### 2. Use delete, trash, restore functions in `PostRepository`

These functions a newly implemented in `PostRepository`, as replacements for the ones in `PostService` which are now deleted in this PR.

#### 3. Show toast messages after trashing posts

This is the UI change part of this PR.

## Test Instructions

Here are a few things that I can think of:
1. Verify the "Undo" button can restore the trashed post as draft.
1. Verify that there is no side-effect in perform "Move to trash" action multiple times on the same post.
1. Verify trashing many posts at the same time works.

## Regression Notes
1. Potential unintended areas of impact
None.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
What's described in the "Test Instructions" section.

3. What automated tests I added (or what prevented me from doing so)
None.

PR submission checklist:


UI Changes testing checklist:
